be1-go/channel/chirp/mod.go

```diff
@@ -0,0 +1,382 @@
+package chirp
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"github.com/rs/zerolog"
+	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/sign/schnorr"
+	"golang.org/x/xerrors"
+	"popstellar/channel"
+	generalChirping "popstellar/channel/generalChirping"
+	"popstellar/channel/inbox"
+	"popstellar/crypto"
+	jsonrpc "popstellar/message"
+	"popstellar/message/answer"
+	"popstellar/message/messagedata"
+	"popstellar/message/query"
+	"popstellar/message/query/method"
+	"popstellar/message/query/method/message"
+	"popstellar/network/socket"
+	"popstellar/validation"
+	"strconv"
+)
+
+const msgID		 = "msg id"
+
+// NewChannel returns a new initialized individual chirping channel
+func NewChannel(channelPath string, ownerKey string, hub channel.HubFunctionalities,
+				generalChannel *generalChirping.Channel, log zerolog.Logger) Channel {
+
+	log = log.With().Str("channel", "chirp").Logger()
+
+	return Channel{
+		sockets:   channel.NewSockets(),
+		inbox:     inbox.NewInbox(channelPath),
+		channelID: channelPath,
+		generalChannel: generalChannel,
+		owner: ownerKey,
+		hub: hub,
+		log: log,
+	}
+}
+
+// Channel is used to handle chirp messages.
+type Channel struct {
+	sockets   channel.Sockets
+	inbox     *inbox.Inbox
+	generalChannel *generalChirping.Channel
+	// channel path
+	channelID string
+	owner string
+	hub channel.HubFunctionalities
+	log zerolog.Logger
+}
+
+
+// Publish is used to handle publish messages in the chirp channel.
+func (c *Channel) Publish(publish method.Publish) error {
+	err := c.verifyPublishMessage(publish)
+	if err != nil {
+		return xerrors.Errorf("failed to verify publish message on a "+
+			"chirping channel: %w", err)
+	}
+
+	msg := publish.Params.Message
+
+	data := msg.Data
+
+	jsonData, err := base64.URLEncoding.DecodeString(data)
+	if err != nil {
+		return xerrors.Errorf("failed to decode message data: %v", err)
+	}
+
+	object, action, err := messagedata.GetObjectAndAction(jsonData)
+	if err != nil {
+		return xerrors.Errorf("failed to get object and action from message data: %v", err)
+	}
+
+	if object == messagedata.ChirpObject {
+
+		switch action {
+		case messagedata.ChirpActionAdd:
+			err := c.publishAddChirp(msg)
+			if err != nil {
+				return xerrors.Errorf("failed to publish chirp: %v", err)
+			}
+		case messagedata.ChirpActionDelete:
+			err := c.publishDeleteChirp(msg)
+			if err != nil {
+				return xerrors.Errorf("failed to delete chirp: %v", err)
+			}
+		default:
+			return answer.NewInvalidActionError(action)
+		}
+	} else {
+		return xerrors.Errorf("object should be \"chirp\" but is %s", object)
+	}
```

Comment:
Fail fast: first check if the object is NOT messagedata.ChirpObject, and return if true. Like so you avoid indentation in your switch block:

```
if object != messagedata.ChirpObject {
  return xerrors.Errorf("object should be \"chirp\" but is %s", object)
}

switch { ...}